drivers/usbc/tcpc/ucpd_stm32gx.c

```diff
@@ -0,0 +1,1442 @@
+	info->dev = dev;
+
+	/* Initialize the work handler */
+	k_work_init(&info->work, ucpd_alert_handler);
+
+	/* Configure CC change alerts */
+	LL_UCPD_WriteReg(config->ucpd_port, IMR,
+			 UCPD_IMR_TYPECEVT1IE | UCPD_IMR_TYPECEVT2IE);
+	LL_UCPD_WriteReg(config->ucpd_port, ICR,
+			 UCPD_ICR_TYPECEVT1CF | UCPD_ICR_TYPECEVT2CF);
+
+	/* SOP'/SOP'' must be enabled via TCPCI call */
+	data->ucpd_rx_sop_prime_enabled = false;
+
+	stm32gx_ucpd_state_init(dev);
+
+	/* Configure and enable the IRQ */
+	config_tcpc_irq();
+}
+
+/**
+ * @brief Initializes the TCPC
+ *
+ * @return 0 on success
+ * @return -EIO on failure
+ */
+static int ucpd_init(const struct device *dev)
+{
+	const struct tcpc_config *const config = dev->config;
+	uint32_t cfg1;
+	int ret;
+
+	/* The UCPD port is disabled in the LL_UCPD_Init function */
+	ret = LL_UCPD_Init(config->ucpd_port,
+			   (LL_UCPD_InitTypeDef *)&config->ucpd_params);
+
+	if (ret == SUCCESS) {
+		/*
+		 * Set RXORDSETEN field to control which types of ordered sets the PD
+		 * receiver must receive.
+		 */
+		cfg1 = LL_UCPD_ReadReg(config->ucpd_port, CFG1);
+		cfg1 |= LL_UCPD_ORDERSET_SOP | LL_UCPD_ORDERSET_SOP1 |
+			LL_UCPD_ORDERSET_SOP2 | LL_UCPD_ORDERSET_HARDRST;
+		LL_UCPD_WriteReg(config->ucpd_port, CFG1, cfg1);
+
+		/* Enable UCPD port */
+		ucpd_port_enable(dev, true);
+
+		/* Initialize the isr */
+		ucpd_isr_init(dev);
+	}
+
+	return 0;
+}
+
+static const struct tcpc_driver_api driver_api = {
+	.init = ucpd_init,
+	.set_alert_handler_cb = ucpd_set_alert_handler_cb,
+	.get_cc = ucpd_get_cc,
+	.set_rx_enable = ucpd_set_rx_enable,
+	.set_vbus_measure_cb = ucpd_set_vbus_measure_cb,
+	.set_discharge_vbus_cb = ucpd_set_discharge_vbus_cb,
+	.check_vbus_level = ucpd_check_vbus_level,
+	.is_rx_pending_msg = ucpd_is_rx_pending_msg,
+	.receive_data = ucpd_receive_data,
+	.transmit_data = ucpd_transmit_data,
+	.get_vbus = ucpd_get_vbus,
+	.set_discharge_vbus = ucpd_set_discharge_vbus,
+	.set_cc = ucpd_set_cc,
+	.set_roles = ucpd_set_roles,
+	.set_vconn_cb = ucpd_set_vconn_cb,
+	.set_vconn = ucpd_set_vconn,
+	.set_cc_polarity = ucpd_cc_set_polarity,
+	.dump_std_reg = ucpd_dump_std_reg,
+	.set_bist_test_mode = ucpd_set_bist_test_mode,
+	.sop_prime_enable = ucpd_sop_prime_enable,
+};
+
+#define DEV_INST_INIT(n) dev_inst[n] = DEVICE_DT_INST_GET(n);
+static void config_tcpc_irq(void)
+{
+	static const struct device
+	*dev_inst[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)];
+
+	if (irq_config) {
+		/* Only configure IRQ line once */
+		irq_config = false;
+		DT_INST_FOREACH_STATUS_OKAY(DEV_INST_INIT)
+
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    ucpd_isr, dev_inst, 0);
+
+		irq_enable(DT_INST_IRQN(0));
+	}
+}
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 0,
+	     "No compatible STM32 TCPC instance found");
+
+/*
+ * @brief This macro converts the psc_ucpdclk pre-scaler division value
+ *        to a register value. reg_val = log_base_2(v)
+ */
+#define PSC_UCPDCLK_MAP(v) \
```

Comment:
would be nice to introduce a generic log2 function, ref. https://elixir.bootlin.com/linux/latest/source/include/linux/log2.h

Reply:
I used include/math/ilog2.h